Sdx.Data.TreeMapper

katsuji-anzai opened the pull request:
## 概要
・API等のレスポンスで利用しているSdx.Data.TreeJsonなどの特定のTreeデータと、複数のSdx.Db.RecordSetを持ち、Treeデータに関連するRecordを、RecordSetの中から抽出してデータを使用する為のクラスです。

katsuji-anzai committed:
Sdx.Data.Tree.Mapperを追加

・TreeJsonなどのTreeデータと、レコードセットを一つのクラスが持ち、使う為のクラスです。

katsuji-anzai committed:
不要な記述が含まれてしまったので、削除

katsuji-anzai committed:
まだcsproj内に不要な記述があったので削除

katsuji-anzai committed:
No newline at end of fileの解消

katsuji-anzai committed:
テスト中用のコメントの削除とTreeのキー指定が間違っていたので修正。

katsuji-anzai committed:
Itemの持つTreeに対応するレコードが、複数タイプある場合に対応出来なかったので改修

・Dictionaryでキー、レコードを保持し、同じTreeに対して、複数タイプのレコードを持てるようにしました。
・各タイプのレコードは、一つのみなのは変えていません。

katsuji-anzai committed:
汎用クラスにあってはいけない用途の記述があったので修正

katsuji-anzai committed:
Sdx.Data.TreeMapperの動作不良があったので修正

・Dictionaryをnewし忘れていて、データがセット出来なかったので修正
・Sdx.Data.TreeMapper.RecordにCountをする機能を追加。

katsuji-anzai committed:
Merge branch 'feat/treeMapper/multiRecord' into proj/treeMapper

katsuji-anzai committed:
Merge branch 'feat/treeMapper/multiRecordBugFix' into proj/treeMapper

gomo commented on Sdx/Data/TreeMapper/Record.cs:
```diff
@@ -0,0 +1,66 @@
+        private string RecordKey;
+        private Func<T, R, bool> Condition;
+
+        public Record(Tree tree, Dictionary<string, Sdx.Db.RecordSet> recordSetList, Func<T, R, bool> condition)
```
３番目の引数がconditionというのが若干わかりづらい気がした。関数なのに名刺だし・・・

「レコードとツリーを結びつけるもの」って感じだと思うのでtreeRecordMatcherとかいいかも、あるいは単純にmatcherいいかな。

プロパティも含めてMatcherにしましょうか。他のところなら直すほどでもないけど、Sdxなので直しましょうか？

😇

gomo replied:
口頭で相談した件ですけど、こういうインターフェースははどうでしょうか？

```c#
var coupons = new Sdx.Data.TreeMapper.Record<Fzdb.Data.Tree.Coupon>(tree);
coupons.AddRecordSet<Fzdb.Orm.Main.Shop>(
  "shop"
  recordSet,
  (coupon, shop) => coupon.ShopId == shop.ShopId
)
```

あ、でもTypeオブジェクト使ってキャストがうまくできないかな・・・
やるなら下記のような感じになるんだけど、試してもらってもいいですか？

```c#
public void AddRecordSet<R>(string key, Sdx.Db.RecordSet recordSet, Func<T, R, bool> matcher) where R : Sdx.Db.Record
{
    recordSetDictionary[key] = new RecordSetData {
      RecordSet = recordSet,
      Matcher = matcher,
      Type = typeof(R)
    }
}
```

で、キャストはこんな感じ

```c#
private T CreateItem(Tree childTree)
{
    var treeItem = new T();
    treeItem.Tree = childTree;
    foreach(KeyValuePair<string, RecordSetData> pair in recordSetDictionary)
    {
      var data = pair.Value;
      treeItem.AddRecord(
        pair.Key,
        data.RecordSet
          .Select(rec => Convert.ChangeType(rec, data.Type))
          .FirstOrDefault(rec => data.Matcher(treeItem, rec))
      );
    }

    return treeItem;
}
```

Convert.ChangeTypeの返り値の型がobjectなのでだめかな〜だめなら下記も試してもらっていいですか？

```c#
     treeItem.AddRecord(
        pair.Key,
        data.RecordSet
          .Select(rec => (dynamic)Convert.ChangeType(rec, data.Type))
          .FirstOrDefault(rec => data.Matcher(treeItem, rec))
      );
```

ふむ・・・

ダメだったらMatcher内で毎回使うときにキャストして使うか・・・でもよく考えたら、Matcherの中の処理って、カラムを取得して見比べるだけなので、型がRecordでも問題ないかもしれないですね。

```c#
var coupons = new Sdx.Data.TreeMapper.Record<Fzdb.Data.Tree.Coupon>(tree);
coupons.AddRecordSet<Fzdb.Orm.Main.Shop>(
  "shop"
  recordSet,
  (coupon, shop) => coupon.ShopId == shop.GetInt32("shop_id")
)
```

一応上記のキャストを試してみてもらえますか？できても、できなくても、その時点で相談してください。



😇

gomo commented on Sdx/Data/TreeMapper/Record.cs:
```diff
@@ -0,0 +1,66 @@
+        private Tree Tree;
+        private Dictionary<string, Sdx.Db.RecordSet> RecordSetDictionary;
+        private string RecordKey;
+        private Func<T, R, bool> Condition;
```
この辺の変数はすべて小文字始まりですね。プロパティじゃなくて、メンバー変数なので。

😇

gomo commented on Sdx/Data/TreeMapper/Record.cs:
```diff
@@ -0,0 +1,66 @@
+            {
+                yield return CreateItem(childTree);
+            }
+        }
```
これはこっちのメソッドはいらないんじゃないでしょうか？コンストラクタに取り出したやつを渡せばいいだけですし。
そうなると、Treeがリストであることを前提にしたクラスになるので、コンストラクタでIsListみたいなので調べて例外にしちゃうのがいいと思います。

😇

gomo commented on Sdx/Data/TreeMapper/Record/Item.cs:
```diff
@@ -0,0 +1,24 @@
+{
+    public class Item
+    {
+        public Tree Tree { get; set; }
```
このsetはアクセス修飾子をinternalにしたほうがいいですね。internalは同じアセンブリ無い、つまり、dllの中からのみアクセス可能なものです。結構使い所多いので覚えておいてください。

あるいはコンストラクタの引数にしてprivateにするか・・・そっちの方いいかな？

😇

katsuji-anzai committed:
複数レコードセットを、TreeMapperにセット出来るようにした動作の修正。

・複数レコードセットを渡し、Treeに対応したレコードを探す際の条件を、複数渡せるようにしていなかったので修正。
・この変更により、TreeMapper.Record.Itemに単一のデータをセットする為に、TreeMapper.Record内に、内部クラスを置いて、そこに集めたデータを元にItemへセットするデータを選別するように変更になりました。

katsuji-anzai committed:
Merge branch 'feat/treeMapper/multiRecordBugFix' into feat/treeMapper/multiRecord

katsuji-anzai committed:
Merge branch 'feat/treeMapper/multiRecord' into proj/treeMapper

katsuji-anzai commented:
@gomo 
前回レビュー時に出た、
* 変数名が大文字始まりにしてしまっていた件。
* ItemにセットするTree単体に該当するレコードを、レコードセットから取得する為の条件が1つしか渡していなかった件。
* GetItemsの引数無し・引数ありを用意したが、コンストラクタで目的のTree階層を渡しておけば、引数ありは必要ない。という改修の件。

こちらの修正をしました。
再レビューお願いします。

gomo commented on Sdx/Data/TreeMapper/Record.cs:
```diff
@@ -7,60 +7,64 @@
             {
-                yield return CreateItem(childTree);
+                recordSetDictionary = new Dictionary<string,RecordSetData>{};
             }
```
すごい細かいですけど、このrecordSetDictionaryは遅延初期化する必要はないですね。なぜなら、このクラスを使うときにこのディクショナリを使わないことがありえないので。どちらかというと、このメソッドを複数回読んだときIF文の処理分計算量が無駄になるので、遅延初期化しないほうがいいということになります。

まあ、細かいですけどSDXなんで直してもらおうかな。細かいけど、考え方として結構大事なんですよね。「全てに理由がある」という考え方。まあ、もちろん理由のないどっちでもいいことってありますけど、それを考えてみるというのは設計の基本です。

😇

katsuji-anzai replied:
なるほど・・・確かに、宣言する時に作ってしまえばOKですね。
> このメソッドを複数回読んだときIF文の処理分計算量が無駄になる

この考え方も、勉強になります。
今後も、なるべく気をつけて考えるようにします。

gomo commented on Sdx/Data/TreeMapper/Record.cs:
```diff
@@ -7,60 +7,64 @@
-
-        public IEnumerable<Sdx.Data.TreeMapper.Record.Item> GetItems()
+        
+        public class RecordSetData
```
このクラスは公開する必要はなさそう。後悔しないように直して起きましょうか・・・ 🤓 

```c#
private class RecordSetData
```

privateでいけるかな？だめだったらinternalで。

😇

gomo commented on Sdx/Data/TreeMapper/Record.cs:
```diff
@@ -7,60 +7,64 @@
         }
 
-        public IEnumerable<Sdx.Data.TreeMapper.Record.Item> GetItems(string key)
+        public void AddRecordSet(string key, Sdx.Db.RecordSet recordSet, Func<T, Sdx.Db.Record, bool> matcher) 
```
口頭で話した件です。例えば、Item.CouponにShopプロパティを作るとすると

```c#
public Orm.Main.Shop Shop
{
  get
  {
    return (Orm.Main.Shop)GetRecord("shop")
  }
}
```
みたいなメソッドになると思いますが、AddRecordSetするのはクライアントコード書く人で、その人が`shop`以外の名前でAddしてしまうというか、いちいち、名前なんだっけ？って見に行かないとAddできないのが気持ち悪いですね。

プロパティ生やすのやめちゃえば、なんとなく筋が通りますが、ORMのレコードは具象クラスで扱うのが基本なのでできればプロパティ生やしたいですね。取得時にキャストするのも手だとは思いますけど・・・

まず、キーを表現するクラスを作ります。これはアイテムのインナークラスでいいかな。

```c#
public class RecordKey
{
  public string Code{get; private set;}
  public RecordKey(string code)
  {
    Code = code;
  }

  public override int GetHashCode()
  {
    return Code.GetHashCode();
  }
  
  public override bool Equals(object obj)
  {
    return Equals(obj as RecordKey)
  }

  public bool Equals(RecordKey recordKey)
  {
    return recordKey != null && recordKey.Code == this.Code;
  }
}
```
overrideされてるメソッドはDictiornaryのキーにするために必要です。で、Tree.Couponに

```c#
public class Coupon : Sdx.Data.TreeMapper.Record.Item
{
  public static readonly RecordKey ShopKey = new RecordKey("shop")
```

で使うときはこんな感じ

```c#
treeMapper.AddRecordSetData(Coupon.ShopKey, recordSet, (coupon, shop) => coupon.ShopId == shop.GetInt32("shop_id"))
```

別のItemにまたShopがあったときこのキーはEqualを返しちゃうのが気持ち悪いけど、とりあえず、一つのTreeMapperに複数のItemが関連づくことはないので、Coupon内でRecordKey.Codeが重複しなければOKですね。

これを厳密にするなら、RecordKeyを抽象クラスにして、Coupon内に継承たRecordKeyCounpon的なものを作って、GetHashCodeとEqualsでクラスも考慮すれば同じクラス内でしか一致しないことになる・・・（一旦考えます）

Sdx.Data.TreeMapper.Record.Itemにはチェック用のメソッドを用意する

```c#
  public bool HasRecordKey(RecordKey recordKey)
  {
    return this.GetType().GetProperties()
      .Where(prop => prop.IsStatic)
      .Where(prop => prop.PropertyType == typeof(RecordKey))
      .Any(prop => {
        var val = prop.GetValue(null, null) as RecordKey
        return val != null && val.Equals(recordKey)
      })
  }
```

このメソッドは多分動くという次元w自信ないのでここを先に試してください。これができないと独自クラスを作る意味がないので。

Stringを使わずわざわざキー用のクラスを作るのは下記の理由です。

* AddRecordSetDataで型チェックが可能
* Itemにチェック用メソッドを作るのに、プロパティとは持っているRecordKeyとは別に配列を宣言しなくて済む。

katsuji-anzai committed:
TreeMapper.Itemのレコード用キーの仕組みを変更

・レコードセットのキーを今まで文字列で「shop」というように引数に渡して、レコードを取得する際にも「shop」とキーを直書きで指定することになっていたのを、Item内部クラスのRecordKeyを使って内部で引き回し、キーの指定間違いをしないで済むように変更しました。

katsuji-anzai committed:
Merge branch 'feat/treeMapper/multiRecord' into proj/treeMapper

katsuji-anzai committed:
RecordKeyのチェックHasRecordKeyの中の挙動が間違っていて、常にFalseになっていたので修正

katsuji-anzai committed:
Merge branch 'feat/treeMapper/multiRecord' into proj/treeMapper

katsuji-anzai commented:
@gomo 

再レビューで出た下記部分修正しましたので、確認お願いします。
* レコードセットと対になるキー指定をする際、内部クラスRecordKeyを引き回す仕様に変更
* Item.Couponの名前空間が微妙だったので、Fzdb.Data.TreeMapper.Record.Item.Couponに修正。

以上です。
よろしくお願いします。

gomo commented on Sdx/Data/TreeMapper/Record/Item.cs:
```diff
@@ -8,19 +8,56 @@ namespace Sdx.Data.TreeMapper.Record
+          if(!HasRecordKey(key))
           {
-            treeRecords = new Dictionary<string, Db.Record> { };
+            throw new ArgumentException("Argument 'key' is invalid. It must be of type 'RecordKey'.");
```
ちょっと前のミーティングで共有したけど、これはエラーメッセージが不親切というか不正確ですね。

まず、ここに来てるのに、typeがRecordKeyじゃないことはないので`It must be of type 'RecordKey'.`というのは？？となります。あと話がそれるけど英文的に変かな？（私もあまり得意じゃないので言うのもなんですが）

`It must be RecordKey type.`

の方がしっくりくるかも。Google翻訳を活用すると作文の練習になるので活用してください。最近コツがわかって来たのですが、主語や目的語を省略しないでしっかり日本語を入力すると割と正確な英語が出ると思います。あと、自分の知らない単語は極力知ってる簡単な単語に置き換えるといいです。

あと、これは「引数に指定したRecordKeyがこのアイテムのプロパティに設定してないですよ〜」というエラーなので、

```c#
throw new ArgumentException("Missing `" + key.Code + "` RecordKey property in this item.");
```

とか、意味をそんな感じにするのと、何よりも何が与えられてるのか`key.Code`で出してあげた方がいいですね。一番大事なのはエラーメッセージに`key.Code`を含んであげることです。

😇

katsuji-anzai replied:
書いていて、このエラー文、微妙・・・と思ったのですが。
なんていう説明をするのが良いのかが、どうしても思いつかなくて、
そもそも、ここで返すエラーは、
「RecordKeyとして認められているものを渡してないですよ」
ではなく、
「RecordKeyの"key.Code"の指定が間違っていますよ」
なんですね。

言われてみると分かるんですが・・・。

ここでエラーを返す意図を、はっきり把握してない状態だったから、うまいエラーメッセージが導き出せなかったのかなぁ・・・って思います。

gomo commented on Sdx/Data/TreeMapper/Record/Item.cs:
```diff
@@ -8,19 +8,56 @@ namespace Sdx.Data.TreeMapper.Record
+            .Where(prop => prop.FieldType == typeof(RecordKey))
+            .Any(prop => {
+              var val = prop.GetValue(null) as RecordKey;
+              return val != null && val.Equals(recordKey);
```
Equalsの実装を見ると

```c#
.Any(prop => recordKey. Equals(prop.GetValue(null)))
```
で問題なさそうだね。ただ、recordKeyがNULLがくるとこけるので、防御してやる必要があるけど・・・

```c#
public bool HasRecordKey(RecordKey recordKey)
{
    if(recordKey == null)
    {
      throw new ....
    }
```

としてやるのがいいかな。return falseどどっちがいいかな・・・基本的にNULLは想定してないと思うので例外がいいかな。

😇

katsuji-anzai committed:
例外のメッセージが不親切な文だったので修正。

katsuji-anzai committed:
Merge branch 'feat/treeMapper/multiRecord' into proj/treeMapper

katsuji-anzai committed:
例外のメッセージを変更

katsuji-anzai commented:
下記修正しました。
・例外の文章を、明確に。
・HasRecordKey内で、recordKeyがnullの時は例外を返すように。

再レビューお願いします。

gomo reviewed (approved):
👍

sunrise-matsumoto committed:
Merge branch 'master' into proj/treeMapper

sunrise-matsumoto committed:
Merge branch 'master' into proj/treeMapper

sunrise-matsumoto commented:
# 本アップ完了しました
